typed-builder-macro/src/enum_info.rs

```diff
@@ -0,0 +1,158 @@
+        let builder_method_visibility = internal_struct_info.builder_method_visibility();
+        let builder_method_name = internal_struct_info
+            .builder_attr
+            .builder_method
+            .get_name()
+            .unwrap_or(syn::Ident::new(&variant_name.to_string().to_case(Case::Snake), Span::call_site()).to_token_stream());
+        let internal_struct_doc_and_visibility = if internal_struct_info.builder_attr.doc {
+            let doc = format!(
+                "
+                Internal struct for building [`{enum_name}::{variant_name}`] instances.
+
+                See [`{enum_name}::{builder_method_name}()`] for more info.
+                ",
+                enum_name = enum_name,
+                variant_name = variant_name,
+                builder_method_name = builder_method_name,
+            );
+            let vis = &self.ast.vis;
+            quote!(#[doc = #doc] #vis)
+        } else {
+            quote!(#[doc(hidden)])
+        };
+        let builder_method_doc = format!(
+            "
+            Create a builder for building [`{enum_name}::{variant_name}`].
+            On the builder, call {setters} to set the values of the fields.
+            Finally, call `.{build_method_name}()` to create the instance of `{enum_name}`.
+            ",
+            enum_name = enum_name,
+            variant_name = variant_name,
+            setters = internal_struct_info.builder_method_setters_doc(),
+            build_method_name = build_method_name,
+        );
+        let internal_struct_derived_tokenstream = internal_struct_info.derive()?;
+        let variant_field_names = variant_fields
+            .named
+            .iter()
+            .map(|f| f.ident.to_token_stream())
+            .collect::<Punctuated<_, Token![,]>>();
+        let variant_field_name_and_types = variant_fields
+            .named
+            .iter()
+            .map(|f| {
+                let (field_name, field_type) = (&f.ident, &f.ty);
+                quote! { #field_name: #field_type, }
+            })
+            .collect::<TokenStream>();
+        let internal_builder_name = &internal_struct_info.builder_name;
+        let internal_builder_method_name = internal_struct_info.builder_method_name();
+        Ok(quote! {
+            #[allow(dead_code, non_camel_case_types, missing_docs)]
+            #internal_struct_doc_and_visibility
+            struct #internal_struct_name { #variant_field_name_and_types }
+
+            #internal_struct_derived_tokenstream
+
+            impl #enum_name {
+                #[doc = #builder_method_doc]
+                #[allow(dead_code)]
+                #builder_method_visibility fn #builder_method_name() -> #internal_builder_name {
+                    #internal_struct_name::#internal_builder_method_name()
+                }
+            }
+
+            #[automatically_derived]
+            impl From<#internal_struct_name> for #enum_name {
```

Comment:
I don't like that we define a potentially visible internal struct just for this. I'd rather modify `DeriveInput` to have an `Option`al "parent enum" field, and have it directly generate the enum variant of that field.